distributed/worker.py

```diff
@@ -2174,20 +2175,25 @@ def memory_monitor(self):
         total = 0
         proc = psutil.Process()
         frac = proc.memory_info().rss / self.memory_limit
+        if frac > min(self.memory_pause_fraction, self.memory_spill_fraction):
+            # Force a GC event so as to get more accurate memory usage
+            # measurements in case GC is lagging.
+            gc.collect()
```

Comment:
How often is `memory_monitor` executed? `gc.collect` can become quite costly if there are many live objects.

Reply:
`gc.collect` is only executed when we are above the high memory fraction. This should not incur any significant performance impact when the memory usage is normal (lower than 70% of the limit by default I think).

Reply:
But imagine a worker that's permanently above that fraction. Now you're running the GC over and again (even if it doesn't release any memory)...

Reply:
Here is the kind of log I got when running the linked snippet with the 1GB memory limit on the worker with additional logs before and after GC:

https://gist.github.com/ogrisel/6cb75c91d515601c4a659675506ea9ff

I had the code running for 10min or so. So it would not get triggered too often but saved a spurious restart.

Reply:
> But imagine a worker that's permanently above that fraction. Now you're running the GC over and again (even if it doesn't release any memory)...

The default `memory_monitor_interval` is 200ms according to the default value of the kwarg. I could add a 5s cool-down period between two consecutive gc calls but that will make the code more complex / hackish.

Reply:
Well, It wouldn't be extraordinary for `gc.collect` alone to take 200 ms. Running it on such a short period doesn't sound like a good idea.

Also, I wonder why `gc.collect` is supposed to fix your memory issue at all. Do you have cyclic references somewhere?

Reply:
It seems to reduce the variance of the measurement of the memory usage of the worker and avoid triggering a worker restart when it's not necessary. If the worker get restarted, all the intermediate tasks are canceled and restarted from scratch which can be very expensive and can even cause the whole "macro-task" to never complete.

Reply:
We should at least call `gc.collect()` when first transitioning to a paused state.  This occurs below after the `if not self.paused` statement.  I'm curious if this suffices to avoid your problem.

Reply:
We may also want to call gc.collect() regularly when memory is use is high, I agree with @pitrou though that we should probably control this a little bit.  @bluenote10 's solution in https://github.com/dask/distributed/pull/1255 might be worth looking into.

Reply:
Calling `gc.collect()` in the paused state is apparently not enough. nanny has still a high probability of reaching it's 95% threshold without the worker having the time to see the internal monitory 80% paused threshold first.

I have tried to make nanny less trigger happy by adding a patience counter for consecutive violations but then the nanny leak test tend to fail with my computer close to freezing for a couple of seconds: the nanny OOM killer became too slow to detect the memory leak.

Reply:
We might call `gc.collect()` *if* we haven't called it fairly recently?  Taking the implementation from #1255 :

```python
class ThrottledGC(object):
    """
    Allows to run throttled garbage collection as a countermeasure to e.g.:
    https://github.com/dask/zict/issues/19
    """
    def __init__(self, min_interval_in_sec=1.0):
        self.min_interval_in_sec = min_interval_in_sec
        self.last_collect = time()
  
    def collect(self, force_gc=False):
        new_time = time()
        if force_gc or new_time - self.last_collect > self.min_interval_in_sec:
            gc.collect()
            self.last_collect = new_time
```

Reply:
I have a tried a less aggressive / smarter strategy with extensive logs here: f3981a53a81af4ddfef97f8ed58a9cd61e3dd76c . It kind of works but I still get a random OOM after a while.

Reply:
At least that last option gets rid of the following erroneous: warning I get most often when running the original snippet on master: "Memory use is high but worker has no data to store to disk.  Perhaps some other process leaking memory?"

Reply:
Nanny OOMs aren't entirely avoidable if your tasks can quickly generate a moderate fraction of your available memory.  It will always be possible that they can conspire to generate enough data to send you over very quickly.

There are costs and benefits to both approaches.

Reply:
I will implement the `ThrottledGC` in this PR. But I am not sure if should better implement the strategy of this PR or the one in f3981a5.

Reply:
I don't have any thoughts on which strategy if better.  If you think what you have in https://github.com/dask/distributed/commit/f3981a53a81af4ddfef97f8ed58a9cd61e3dd76c is nicer then go with that.